### PR TITLE
Implement ATOM shader instruction

### PR DESCRIPTION
**Pull request by gdkchan**
This implements the ATOM shader instruction, and fixes negative offsets on the ATOMS shader instructions (the offset was not being sign-extended).

The ATOM instruction is similar to ATOMS, but operates on global memory. It can be used to do atomic operations on memory, such as GLSL `atomicAdd` and friends.

I know a bunch of games that uses this (Fire Emblem Three Houses is one of them), but I don't know any specific issue that this change fixes.

Note: The compare and swap variant is still not implemented, but it also doesn't seems to be implemented for the other atomic instructions.

### CODE REVIEW

**Review by riperiperi — APPROVED** [review]
👍

File: Ryujinx.Graphics.Shader/Instructions/InstEmitMemory.cs [hunk 1]
```diff
@@ -57,13 +57,37 @@ public static void Ast(EmitterContext context)
...
+
+            (Operand addrLow, Operand addrHigh) = Get40BitsAddress(context, op.Ra, op.Extended, sOffset);
+
+            Operand value = GetSrcB(context);
+
+            Operand res = EmitAtomicOp(
+                context,
+                Instruction.MrGlobal,
+                op.AtomicOp,
+                op.Type,
```
Comment (Thog) [inline Ryujinx.Graphics.Shader/Instructions/InstEmitMemory.cs:74]: Wrong, this is reusing bits 28 to 30 here, https://github.com/Ryujinx/Ryujinx/blob/master/Ryujinx.Graphics.Shader/Decoders/OpCodeAtom.cs#L27

the type is stored starting at bit 49 with the given possible values for ATOM (opcode: 0xed)

```cs
enum AtomType
{
        U32 = 0,
        S32 = 1,
        U64 = 2,
        F32 = 3,
}

File: Ryujinx.Graphics.Shader/Instructions/InstEmitMemory.cs [hunk 2]
```diff
@@ -57,13 +57,37 @@ public static void Ast(EmitterContext context)
...
+            int sOffset = (op.RawOpCode.Extract(28, 20) << 12) >> 12;
+
+            (Operand addrLow, Operand addrHigh) = Get40BitsAddress(context, op.Ra, op.Extended, sOffset);
+
+            Operand value = GetSrcB(context);
+
+            Operand res = EmitAtomicOp(
+                context,
+                Instruction.MrGlobal,
+                op.AtomicOp,
```
Comment (Thog) [inline Ryujinx.Graphics.Shader/Instructions/InstEmitMemory.cs:73]: NOTE here:  ATOM (opcode: 0xed) also have a possible "safe add" (0b1010) operation, would be nice to add it to AtomicOp and add a comment about it. (only valid for this opcode)

**Review by Thog — CHANGES_REQUESTED** [review]
Need changes

File: Ryujinx.Graphics.Shader/Instructions/InstEmitMemory.cs (hunk 1 above)
Comment (gdkchan) [reply Ryujinx.Graphics.Shader/Instructions/InstEmitMemory.cs:74]: Nice catch, I will fix it.

**Review by Thog — APPROVED** [review]

**Pull request merged**